JAVA-3069: pinning mongos in sharded transactions

jstewart148 opened the pull request:
I wanted to submit this PR for review while trying to figure out how to run tests with multiple mongos on Evergreen. Both the sync and async tests I have added work locally with a sharded cluster with multiple mongos instances using the latest version of MongoDB.

I will add an Evergreen patch to this PR soon.

jstewart148 committed:
JAVA-3069: support mongos pinning for sharded txns

jstewart148 committed:
JAVA-3069: mongos pinning for sharded transactions

jstewart148 committed:
JAVA-3069: fix checkstyle errors

jyemin committed:
Improve CallbackResultHolder exception message to aid debugging

JAVA-3038

jyemin committed:
Add earlier check for failure in RetryableWritesTest

JAVA-3038

jstewart148 committed:
JAVA-3069: async pinning fixes

jstewart148 committed:
JAVA-3069: experiment with run-tests script

jyemin committed:
Vendor https://github.com/marianobarrios/tls-channel@0.2.0

Modified to conform to driver style and compile at Java 6 language level

JAVA-3038

jyemin committed:
Add TlsChannel stream support

JAVA-3038

jyemin committed:
Add connectivity test for async

JAVA-3038

jyemin committed:
Don't require both trust and key store configuration when running tests

jyemin committed:
Default to TlsChannelStreamFactoryFactory in async

when running on Java 8+ and SSL is enabled

JAVA-3038

jyemin committed:
In tests, use TlsChannel when possible, otherwise use Netty

JAVA-3038

jyemin committed:
Refactor AsynchronousSocketChannelStream

AsynchronousSocketChannelStream now extends common base class
AsynchronousChannelStream along with TlsChannelStream.

JAVA-3038

jstewart148 committed:
JAVA-3069: use single server binding for pinning mongos

jstewart148 committed:
JAVA-3069: alter run-tests.sh

jstewart148 committed:
JAVA-3069: update run-tests script to use all mongos instances

jstewart148 committed:
JAVA-3069: update run-tests.sh

jstewart148 committed:
JAVA-3069: fix checkstyle errors

jstewart148 committed:
JAVA-3069: update run-tests.sh

jstewart148 committed:
JAVA-3069: checkstyle fixes

jstewart148 committed:
JAVA-3069: fix bug in MongoClientDelegate.getReadWriteBinding()

jstewart148 committed:
JAVA-3069: fix groovy test

jstewart148 committed:
JAVA-3069: restore original code in run-tests.sh

jstewart148 committed:
JAVA-3069: fix groovy code violation

jstewart148 committed:
JAVA-3069: fix leak of client sessions

jstewart-mongo commented:
I fixed the session leak. Tests run clean locally and results are looking better on Evergreen.

New Evergreen patch: https://evergreen.mongodb.com/version/5c3f534957e85a4a0ae177d0

jstewart148 committed:
JAVA-3069: remove connection string for sharded cluster in run-tests.sh

jstewart148 committed:
JAVA-3069: add Java system property for transaction test

jyemin commented on driver-core/src/main/com/mongodb/internal/session/BaseClientSessionImpl.java:
```diff
@@ -33,16 +35,29 @@
+
+    @Override
+    public void setPinnedMongosAddress(@Nullable final ServerAddress address) {
+        pinnedMongosAddress = address;
```
To be safe, this should throw an exception if pinnedMongoAddress is not currently null.

jstewart148 replied:
`pinnedMongosAddress` can be set to null when a transaction is started.

jyemin replied:
Yea, that's true.  So it would have to be something like:

   isTrue(address == null || pinnedMongosAddress == null)

jstewart148 replied:
Done

jyemin commented on driver-sync/src/main/com/mongodb/client/internal/ClientSessionImpl.java:
```diff
@@ -42,11 +44,24 @@
     private boolean messageSentInCurrentTransaction;
     private boolean commitInProgress;
     private TransactionOptions transactionOptions;
+    private ServerAddress pinnedMongosAddress;
```
Seems like there are two identical fields, one here and one in the base class.  That shouldn't be necessary.

jstewart148 replied:
Removed the field and refer to the field in the base class.

jyemin commented on driver-sync/src/main/com/mongodb/client/internal/MongoClientDelegate.java:
```diff
@@ -210,8 +216,21 @@ ReadBinding getReadBinding(final ReadPreference readPreference, final ReadConcer
-                    readConcern);
+            ReadWriteBinding readWriteBinding;
+            if (session != null && session.hasActiveTransaction()
+                    && cluster.getCurrentDescription().getType() == ClusterType.SHARDED) {
```
I don't think this is safe.  Cluster#getCurrentDescription() doesn't block waiting to discover the cluster type (e.g. it can return ClusterType.UNKNOWN), so this could fail to pin if the very first thing you do after creating the MongoClient is to start a transaction.

I think you can just replace this with a call to Cluster#getDescription.  Hopefully I didn't have a reason for writing it this way in the code I pasted into the dev channel.  Maybe it'll come to me but for now I think this is safer.

jstewart148 replied:
Replaced the call with `Cluster#getDescription`.

jyemin commented on driver-core/src/main/com/mongodb/connection/ServerType.java:
```diff
@@ -83,9 +83,9 @@ public ClusterType getClusterType() {
+     * A mongos server.
      */
-    SHARD_ROUTER {
+    MONGOS {
```
This is a backwards breaking change.  It has to be reverted.

jstewart148 replied:
Done

jyemin commented on driver-core/src/main/com/mongodb/connection/ServerDescription.java:
```diff
@@ -501,21 +501,21 @@ public ServerAddress getAddress() {
      */
-    public boolean isReplicaSetMember() {
-        return type.getClusterType() == ClusterType.REPLICA_SET;
+    public boolean isMongos() {
```
This is a backwards breaking change.  It has to be reverted.

jstewart148 replied:
Done

jyemin commented on driver-core/src/main/com/mongodb/session/ClientSession.java:
```diff
@@ -33,6 +35,23 @@
+     * Get the server address of the pinned mongos on this session.
+     *
+     * @return the server address of the pinned mongos
+     * @mongodb.server.release 4.0
```
Should be 4.2, right?

jstewart148 replied:
I've made this change.

jyemin commented on driver-core/src/main/com/mongodb/session/ClientSession.java:
```diff
@@ -33,6 +35,23 @@
 public interface ClientSession extends Closeable {
 
+    /**
+     * Get the server address of the pinned mongos on this session.
```
I think it's worth a comment saying this is for internal use only.

jstewart148 replied:
Done.

jstewart148 replied:
Comment added.

jyemin commented on driver-core/src/main/com/mongodb/session/ClientSession.java:
```diff
@@ -33,6 +35,23 @@
+    @Nullable
+    ServerAddress getPinnedMongosAddress();
+
+    /**
```
As above

jstewart148 replied:
Done

jstewart148 replied:
Comment added.

jyemin commented on driver-core/src/main/com/mongodb/binding/AsyncSingleServerBinding.java:
```diff
@@ -0,0 +1,143 @@
+/**
+ * A simple binding where all connection sources are bound to the server specified in the constructor.
+ *
+ * @since 3.0
```
Nit: since 3.11

jstewart148 replied:
Changed.

jyemin reviewed (changes requested):
Initial feedback included here.  It would be helpful if you revert the backwards breaking changes I pointed out, as it will dramatically reduce the number of changed files that we need to look at.

Also, I was expecting to see the updated transaction specification tests in this review.  Am I mistaken that there are new tests?

** Correction: I do see the one new test: pin-mongos.json.

jyemin commented on driver-sync/src/test/functional/com/mongodb/client/PinMongosTest.java:
```diff
@@ -0,0 +1,302 @@
+import static org.junit.Assume.assumeTrue;
+
+@RunWith(Parameterized.class)
+public class PinMongosTest extends DatabaseTestCase {
```
I'm not clear why we need a separate test runner for the one new set of tests.  Is the existing TransactionsTest not sufficient?  If it's not, let's figure out how to make it so it is rather than having to add another test runner.

jstewart148 replied:
I was able to combine the tests.

jyemin commented on driver-core/src/test/resources/sharded-transactions/pin-mongos.json:
```diff
@@ -0,0 +1,794 @@
+{
```
Can we just add this file into the src/test/resources/transactions directory?

jstewart148 replied:
Done

jstewart148 committed:
JAVA-3069: code review changes

jstewart148 committed:
JAVA-3069: add test to TransactionTest

jstewart-mongo commented:
Latest Evergreen patch: https://evergreen.mongodb.com/version/5c3ff57b1e2d17417504aa52

jstewart148 committed:
JAVA-3069: add transaction test Java property

jstewart-mongo commented:
Evergreen patch to test transaction Java property: https://evergreen.mongodb.com/version/5c401594d1fe073ed3f47584

jstewart148 committed:
JAVA-3069: fix run-tests.sh script

rozza committed:
Added more driver name information

Previously, all drivers (legacy, sync, async) would just report the driver name "mongo-java-driver".
This update appends extra information to the driver name:

* legacy    : mongo-java-driver|legacy
* sync      : mongo-java-driver|sync
* async     : mongo-java-driver|async
* netty     : mongo-java-driver|async|netty

JAVA-3131

jyemin commented on .evergreen/run-tests.sh:
```diff
@@ -84,13 +85,20 @@ fi
    provision_ssl
 fi
+
+if [ $SAFE_FOR_MULTI_MONGOS == true ]; then
```
If all tests now pass with multiple mongos in the connection string on 4.2+, then there is no need for `org.mongodb.test.transaction.uri`.  That was our backstop in case all tests didn't pass

jstewart148 replied:
The Evergreen patch run with this change was a false positive since the property was not properly passed to .evg.yml.

jyemin commented on .evergreen/.evg.yml:
```diff
@@ -469,6 +469,7 @@ axes:
         display_name: "latest"
         variables:
            VERSION: "latest"
+           SAFE_FOR_MULTI_MONGOS: true
```
This deserves an explanatory comment

jstewart148 replied:
Comment added

jyemin commented on driver-async/src/main/com/mongodb/async/client/ClientSessionImpl.java:
```diff
@@ -41,14 +43,26 @@
     private boolean commitInProgress;
-
     private TransactionOptions transactionOptions;
+    private ServerAddress pinnedMongosAddress;
```
As with the sync version, you can remove this and delegate it to the base class

jstewart148 replied:
Done

jyemin commented on driver-async/src/main/com/mongodb/async/client/ClientSessionImpl.java:
```diff
@@ -41,14 +43,26 @@
+        this.pinnedMongosAddress = null;
+    }
+
+    @Override
```
These methods can be removed since the base class handles it.

jstewart148 replied:
Done

jyemin commented on driver-async/src/main/com/mongodb/async/client/OperationExecutorImpl.java:
```diff
@@ -137,8 +141,22 @@ private void labelException(final Throwable t, final ClientSession session) {
+        AsyncReadWriteBinding readWriteBinding;
+        Cluster cluster = mongoClient.getCluster();
+        if (session != null && session.hasActiveTransaction()
+                && cluster.getCurrentDescription().getType() == ClusterType.SHARDED) {
```
As with the sync version, getCurrentDescription is problematic here.  But the solution is trickier because you can't block in the async driver, and just calling getDescription can be a  blocking operation.  You'll have to use a similar trick to `com.mongodb.async.client.ClientSessionHelper#createClientSession` to do this in a non-blocking fashion.

jstewart148 replied:
I don't understand how `com.mongodb.async.client.ClientSessionHelper#createClientSession` gets the `ClusterDescription` in a non-blocking fashion. In `createClientSession`, `getCurrentDescription` is called directly:
```
        } else {
            ClusterDescription clusterDescription = mongoClient.getCluster().getCurrentDescription();
            if (!getServerDescriptionListToConsiderForSessionSupport(clusterDescription).isEmpty()
                    && clusterDescription.getLogicalSessionTimeoutMinutes() != null
                    && clusterDescription.getType() != ClusterType.STANDALONE) {
                callback.onResult(createClientSession(options, executor), null);
```
What is the trick here that would be helpful in `getReadWriteBinding` for getting the cluster description?

jyemin commented on driver-core/src/main/com/mongodb/internal/connection/AbstractMultiServerCluster.java:
```diff
@@ -330,7 +330,7 @@ private boolean isNotAlreadyPrimary(final ServerAddress address) {
         if (!newDescription.isShardRouter()) {
             LOGGER.error(format("Expecting a %s, but found a %s.  Removing %s from client view of cluster.",
-                                 SHARD_ROUTER, newDescription.getType(), newDescription.getAddress()));
+                    SHARD_ROUTER, newDescription.getType(), newDescription.getAddress()));
```
Nit: remove white space change so that this file doesn't show up in the diffs.

jstewart148 replied:
Done

jyemin commented on driver-core/src/test/functional/com/mongodb/ClusterFixture.java:
```diff
@@ -89,6 +90,7 @@
 public final class ClusterFixture {
     public static final String DEFAULT_URI = "mongodb://localhost:27017";
     public static final String MONGODB_URI_SYSTEM_PROPERTY_NAME = "org.mongodb.test.uri";
+    public static final String MONGODB_TRANSACTION_URI_SYSTEM_PROPERTY_NAME = "org.mongodb.test.transaction.uri";
```
As mentioned in my comment about run-tests.sh, I don't think this is needed if all tests pass on 4.2+ with multiple mongos.  In that case, all tests can use the same connection string.

jstewart148 replied:
The Evergreen patch run after this code change produced a false positive. Using this property for transaction tests only.

jyemin commented on .evergreen/run-tests.sh:
```diff
@@ -65,12 +65,13 @@ provision_ssl () {
 # Provision the correct connection string and set up SSL if needed
 if [ "$TOPOLOGY" == "sharded_cluster" ]; then
-
+    if [ $SAFE_FOR_MULTI_MONGOS != true ]; then
```
I think you have to pass this in as an argument to the shell script where it's called from .evg.yml, as with all the other variables.

jstewart148 replied:
Now passing a TRANSACTION_URI variable to .evg.yml.

rozza committed:
Revert driver metadata for netty

It doesn't catch programmatically set netty streamfactories. Also
as StreamType is being removed in 4.0 along with NettyMongoClients.java

JAVA-3131

jstewart148 committed:
JAVA-3069: remove Java property for multi mongos

rozza committed:
Temporarily disable android test

JAVA-3142

jstewart148 committed:
JAVA-3069: add Java system property for transactions

jstewart148 committed:
JAVA-3069: checkstyle fixes

jstewart148 committed:
JAVA-3069: async changes in OperationExecutorImpl

rozza commented on driver-async/src/main/com/mongodb/async/client/OperationExecutorImpl.java:
```diff
@@ -69,22 +78,28 @@ public void onResult(final ClientSession clientSession, final Throwable t) {
+                                public void onResult(final AsyncReadWriteBinding binding, final Throwable t) {
+                                    if (session != null && session.hasActiveTransaction()
+                                            && !binding.getReadPreference().equals(primary())) {
+                                        throw new MongoClientException("Read preference in a transaction must be primary");
```
You can't throw in callback - you must pass the exception up the callback stack.

jstewart148 committed:
JAVA-3069: fix getReadWriteBinding

jyemin commented on driver-async/src/main/com/mongodb/async/client/OperationExecutorImpl.java:
```diff
@@ -69,22 +78,38 @@ public void onResult(final ClientSession clientSession, final Throwable t) {
+                                public void onResult(final AsyncReadWriteBinding binding, final Throwable t) {
+                                    if (t != null) {
+                                        errHandlingCallback.onResult(null, t);
+                                        if (binding != null) {
```
Generally you can assume that if t != null, then result == null.  Similar to they way a synchronous method can either return normally or throw..

jyemin commented on driver-async/src/main/com/mongodb/async/client/OperationExecutorImpl.java:
```diff
@@ -69,22 +78,38 @@ public void onResult(final ClientSession clientSession, final Throwable t) {
+                                    } else {
+                                        if (session != null && session.hasActiveTransaction()
+                                                && !binding.getReadPreference().equals(primary())) {
+                                            binding.release();
```
Looks like this fixes a bug.  In the previous version we wouldn't release the binding before we throw.  Nice.

jyemin commented on driver-async/src/main/com/mongodb/async/client/OperationExecutorImpl.java:
```diff
@@ -107,25 +132,36 @@ public void onResult(final ClientSession clientSession, final Throwable t) {
+                                public void onResult(final AsyncReadWriteBinding binding, final Throwable t) {
+                                    if (t != null) {
+                                        errHandlingCallback.onResult(null, t);
+                                        if (binding != null) {
```
As above.  You can assume binding == null.

rozza committed:
Reenabled android tests and moved ExtendedAsynchronousByteChannel

Moved ExtendedAsynchronousByteChannel to com.mongodb.internal.connection
The dex'ing process didn't like it being referenced in a higher package
when the code was split across dex files.

JAVA-3142

jyemin commented on driver-async/src/main/com/mongodb/async/client/OperationExecutorImpl.java:
```diff
@@ -134,15 +170,92 @@ private void labelException(final Throwable t, final ClientSession session) {
+    }
+
+    private void getClusterDescription(final Cluster cluster, final SingleResultCallback<ClusterType> callback) {
+        ClusterDescription description = cluster.getDescription();
```
This should be cluster.getCurrentDescription(), since cluster.getDescription() can block.

jyemin reviewed (changes requested):
Almost there.  Just a few remaining things, mostly cosmetic.

jstewart148 committed:
JAVA-3069: code review changes

jstewart148 committed:
JAVA-3069: checkstyle changes

jstewart-mongo commented:
Resolved the callback issues. `./gradlew check` succeeds. 
Evergreen patch: https://evergreen.mongodb.com/version/5c42204e32f4176719ca6a86

jstewart148 committed:
JAVA-3069: checkstyle fix

jstewart148 committed:
JAVA-3069: fix transaction tests

jstewart-mongo commented:
Evergreen patch: https://evergreen.mongodb.com/version/5c42576b57e85a6afe064c2e

jyemin commented on driver-async/src/test/functional/com/mongodb/async/client/TransactionsTest.java:
```diff
@@ -130,7 +135,22 @@ public void setUp() {
-        mongoClient = MongoClients.create(Fixture.getMongoClientBuilderFromConnectionString()
+        ConnectionString connectionString = getConnectionString();
+        if (this.filename.equals("pin-mongos.java")) {
+            connectionString = getMultiMongosConnectionString();
```
Do tests all pass if we use the multiMongos connection string for all tests?  They should, and I think we'd get slightly better test coverage if we did use that, since as we saw the pin-mongos tests don't have a lot of assertions in them.

jyemin replied:
Ping on this one.

jstewart148 replied:
No, we do not pass all tests. When running the tests with the multiMongos connection string locally, 64 of the 144 tests failed. Many of the tests failed with an error like
```
com.mongodb.MongoCommandException: Command failed with error 263 (OperationNotSupportedInTransaction): 'Encountered error from localhost:27020 during a transaction :: caused by :: Cannot create namespace JavaDriverTest.test in multi-document transaction.' on server localhost:27037. The full response is {"ok": 0.0, "errmsg": "Encountered error from localhost:27020 during a transaction :: caused by :: Cannot create namespace JavaDriverTest.test in multi-document transaction.", "code": 263, "codeName": "OperationNotSupportedInTransaction", "operationTime": {"$timestamp": {"t": 1547864149, "i": 25}}, "$clusterTime": {"clusterTime": {"$timestamp": {"t": 1547864149, "i": 25}}, "signature": {"hash": {"$binary": "AAAAAAAAAAAAAAAAAAAAAAAAAAA=", "$type": "00"}, "keyId": {"$numberLong": "0"}}}}
```

jstewart148 replied:
Other tests produce results that do not match the expected results.

jyemin replied:
OK, thanks for checking.

jyemin commented on driver-async/src/test/functional/com/mongodb/async/client/TransactionsTest.java:
```diff
@@ -130,7 +135,22 @@ public void setUp() {
 
-        mongoClient = MongoClients.create(Fixture.getMongoClientBuilderFromConnectionString()
+        ConnectionString connectionString = getConnectionString();
+        if (this.filename.equals("pin-mongos.java")) {
```
This should be "pin-mongos.json"

jyemin replied:
But maybe irrelevant given my comment below.

jstewart148 committed:
JAVA-3069: fix transaction test for sharded cluster

jstewart148 committed:
JAVA-3069: massage actual resutls in transaction tests
